manifest/dynamic_manifest.go

```diff
@@ -31,169 +35,156 @@ type DynamicManifestProvider struct {
-	return &DynamicManifestProvider{
-		manifest:         manifest,
-		pubsub:           pubsub,
-		ec:               ec,
-		manifestServerID: manifestServerID,
-	}
+	initialManifest *Manifest
+	manifestChanges chan *Manifest
 }
 
-func (m *DynamicManifestProvider) Manifest() Manifest {
-	m.lk.RLock()
-	defer m.lk.RUnlock()
-	return m.manifest
+type ManifestUpdateMessage struct {
+	MessageSequence uint64
+	ManifestVersion uint64
+	Manifest        *Manifest
 }
 
-func (m *DynamicManifestProvider) GpbftOptions() []gpbft.Option {
-	return m.manifest.GpbftOptions()
-}
+func (mu ManifestUpdateMessage) toManifest() *Manifest {
+	if mu.Manifest == nil {
+		return nil
+	}
 
-func (m *DynamicManifestProvider) SetManifestChangeCallback(mc OnManifestChange) {
-	m.lk.Lock()
-	defer m.lk.Unlock()
-	m.onManifestChange = mc
+	// When a manifest configuration changes, a new network name is set that depends on the
+	// manifest version of the previous version to avoid overlapping previous configurations.
+	cpy := *mu.Manifest
+	newName := fmt.Sprintf("%s/%d", string(cpy.NetworkName), mu.ManifestVersion)
+	cpy.NetworkName = gpbft.NetworkName(newName)
+	return &cpy
 }
 
-// When a manifest configuration changes, a new network name
-// is set that depends on the manifest version of the previous version to avoid
-// overlapping previous configurations.
-func (m *DynamicManifestProvider) networkNameOnChange() gpbft.NetworkName {
-	return gpbft.NetworkName(string(m.manifest.NetworkName) + "/" + fmt.Sprintf("%d", m.manifest.Sequence))
+func (m ManifestUpdateMessage) Marshal() ([]byte, error) {
+	b, err := json.Marshal(m)
+	if err != nil {
+		return nil, xerrors.Errorf("marshaling JSON: %w", err)
+	}
+	return b, nil
 }
 
-func (m *DynamicManifestProvider) Run(ctx context.Context, errCh chan error) {
-	if m.onManifestChange == nil {
-		errCh <- xerrors.New("onManifestChange is nil. Callback for manifest change required")
+func (m *ManifestUpdateMessage) Unmarshal(r io.Reader) error {
+	err := json.NewDecoder(r).Decode(&m)
+	if err != nil {
+		return xerrors.Errorf("decoding JSON: %w", err)
 	}
-	go m.handleIncomingManifests(ctx, errCh)
-	m.handleApplyManifest(ctx, errCh)
+	return nil
 }
 
-func (m *DynamicManifestProvider) handleApplyManifest(ctx context.Context, errCh chan error) {
-	// add a timer for EC period
-	ticker := time.NewTicker(ManifestCheckTick)
-	defer ticker.Stop()
-
-	for {
-		select {
-		case <-ticker.C:
-			m.lk.Lock()
-			if m.nextManifest != nil {
-				ts, err := m.ec.GetHead(ctx)
-				if err != nil {
-					log.Errorf("error fetching chain head: %+v", err)
-					m.lk.Unlock()
-					continue
-				}
+func NewDynamicManifestProvider(initialManifest *Manifest, pubsub *pubsub.PubSub, ec ec.Backend, manifestServerID peer.ID) ManifestProvider {
+	ctx, cancel := context.WithCancel(context.Background())
+	errgrp, ctx := errgroup.WithContext(ctx)
 
-				// if the upgrade epoch is reached or already passed.
-				if ts.Epoch() >= m.nextManifest.BootstrapEpoch {
```

Comment:
We now handle this delay with the rest of the bootstrap logic in the `gpbftRunner`.